src/cacheHttpClient.ts

```diff
@@ -83,46 +107,183 @@ async function pipeResponseToStream(
 
-    const cacheUrl = getCacheUrl();
-    const token = process.env["ACTIONS_RUNTIME_TOKEN"] || "";
-    const bearerCredentialHandler = new BearerCredentialHandler(token);
+    const reserveCacheRequest: ReserveCacheRequest = {
+        key
+    };
+    const response = await restClient.create<ReserveCacheResponse>(
+        "caches",
+        reserveCacheRequest,
+        getRequestOptions()
+    );
 
-    const resource = `_apis/artifactcache/cache/${encodeURIComponent(key)}`;
-    const postUrl = cacheUrl + resource;
+    return response?.result?.cacheId ?? -1;
+}
 
-    const restClient = new RestClient("actions/cache", undefined, [
-        bearerCredentialHandler
-    ]);
+function getContentRange(start: number, end: number): string {
+    // Format: `bytes start-end/filesize
+    // start and end are inclusive
+    // filesize can be *
+    // For a 200 byte chunk starting at byte 0:
+    // Content-Range: bytes 0-199/*
+    return `bytes ${start}-${end}/*`;
+}
 
+async function uploadChunk(
+    restClient: RestClient,
+    resourceUrl: string,
+    data: NodeJS.ReadableStream,
+    start: number,
+    end: number
+): Promise<IRestResponse<void>> {
+    core.debug(
+        `Uploading chunk of size ${end -
+            start +
+            1} bytes at offset ${start} with content range: ${getContentRange(
+            start,
+            end
+        )}`
+    );
     const requestOptions = getRequestOptions();
     requestOptions.additionalHeaders = {
-        "Content-Type": "application/octet-stream"
+        "Content-Type": "application/octet-stream",
+        "Content-Range": getContentRange(start, end)
     };
 
-    const response = await restClient.uploadStream<void>(
-        "POST",
-        postUrl,
-        stream,
+    const uploadChunkRequest = async (): Promise<IRestResponse<void>> => {
+        return await restClient.uploadStream<void>(
+            "PATCH",
+            resourceUrl,
+            data,
+            requestOptions
+        );
+    };
+
+    const response = await uploadChunkRequest();
+    if (isSuccessStatusCode(response.statusCode)) {
+        return response;
+    }
+
+    if (isRetryableStatusCode(response.statusCode)) {
+        const retryResponse = await uploadChunkRequest();
+        if (isSuccessStatusCode(retryResponse.statusCode)) {
+            return retryResponse;
+        }
+    }
+
+    throw new Error(
+        `Cache service responded with ${response.statusCode} during chunk upload.`
+    );
+}
+
+async function uploadFile(
+    restClient: RestClient,
+    cacheId: number,
+    archivePath: string
+): Promise<void> {
+    // Upload Chunks
+    const fileSize = fs.statSync(archivePath).size;
+    const resourceUrl = getCacheApiUrl() + "caches/" + cacheId.toString();
+    const responses: IRestResponse<void>[] = [];
+    const fd = fs.openSync(archivePath, "r");
+
+    const concurrency = 4; // # of HTTP requests in parallel
+    const MAX_CHUNK_SIZE = 32000000; // 32 MB Chunks
```

Comment:
might be nice if these were variables to allow overriding/testing easily.

Reply:
Added env variables to control these two settings, with a fallback to the default values